Adjust Dune 2000 missions difficulty

Porenutak opened the pull request:
There was lots of complains that later mission are almost impossible too finish on any difficulty. This is mostly because AI overwhelm you at early game with lots of units when you dont have any money or production established. This will became even more problem when balac patch come out. (T3 units and support powers will become more powerfull) 
This PR try to address this problem by adding:
- InitialProductionDelay for enemy players
- different AttackDelays for early game and late game
- Different unit composition for early game and late game
- Delay using support powers into late game
- adds crush logic for AI

Changed missions:
harkonnen 04 - 09
ordos 05 - 06
Atreides 04 - 05

JovialFeline commented:
I will have to test this later but from past tinkering, I can say harkonnen-05's difficulty is a bit screwy for a reason untouched here. The original mission does not reinforce imperial units to the starport unless the outpost is destroyed. In the ORA script, that reinforcement is done before the check and the outpost status only determines if that reinforcement is announced.

Porenutak commented:
Fixed harkonnen-05 Corrino reinforcements

JovialFeline commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -277,3 +282,43 @@ ProduceUnits = function(player, factory, delay, toBuild, attackSize, attackThres
+	end
+	local actors = Map.ActorsInCircle(unit.CenterPosition, WDist.FromCells(5), function (a)
+		return
+		a.IsDead == false and
+		a.IsInWorld == true and
```
```suggestion
```
ActorsInCircle/ActorsInBox has an IsInWorld check. Is there a case where a dead actor would be in the world?

Porenutak replied:
If its true, they probably not needed. But I cant find any documentation/code mention this. 
What about infantry during death animation? is it already mark as death or not? What about the husks?
I would rather have checks everywhere then dealing with some edge case `isDead `crashes later.

JovialFeline replied:
1. Inflict damage. Repeat [until the victim is considered dead](https://github.com/OpenRA/OpenRA/blob/44c6747decd4d98c2267b64bc7b629d888b9b66d/OpenRA.Mods.Common/Traits/Health.cs#L216) at 0 HP.
2. Notify Killed traits.
    - OnDeathAnimation [prepares a SpriteEffect that will play a sequence as a frame end task](https://github.com/OpenRA/OpenRA/blob/44c6747decd4d98c2267b64bc7b629d888b9b66d/OpenRA.Mods.Common/Traits/Health.cs#L216) (like standard explosions or trails seem to do). Some info from the dead actor is passed along, like the palette.
    - SpawnActorOnDeath [records the actor's killer](https://github.com/OpenRA/OpenRA/blob/44c6747decd4d98c2267b64bc7b629d888b9b66d/OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs#L89) if possible.
3. [Schedule the actor's disposal](https://github.com/OpenRA/OpenRA/blob/44c6747decd4d98c2267b64bc7b629d888b9b66d/OpenRA.Game/Actor.cs#L417) as a frame end task.
4. Set IsInWorld to false and [dispose of the dead actor](https://github.com/OpenRA/OpenRA/blob/44c6747decd4d98c2267b64bc7b629d888b9b66d/OpenRA.Game/World.cs#L357).
5. Notify RemovedFromWorld traits.
    - SpawnActorOnDeath [schedules a husk's creation](https://github.com/OpenRA/OpenRA/blob/44c6747decd4d98c2267b64bc7b629d888b9b66d/OpenRA.Mods.Common/Traits/SpawnActorOnDeath.cs#L143) as a frame end task.
6. SpawnActorOnDeath creates a husk.

After reading and testing, that is the timeline as I understand it. Everything after death occurs on the same world tick, but disposal and post-disposal always take place near the tick's end. It does not seem possible to grab a dead actor from the actors array through ActorsInBox, ActorsInCircle, or ActorsInWorld for the current tick, which is nice. If so, all the checks from Where could probably be put here instead.

JovialFeline commented on mods/d2k/maps/harkonnen-09a/harkonnen09a-AI.lua:
```diff
@@ -46,35 +93,101 @@ ActivateAI = function()
+		ProduceUnits(CorrinoSmall, CLightFactory2, delay, vehilcesToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
+		ProduceUnits(CorrinoSmall, CHeavyFactory2, delay, tanksToBuildCorrinoSmall, AttackGroupSize[Difficulty], attackThresholdSize)
+	end)
+	Trigger.OnProduction(CHeavyFactory2, function (producer, produced)
+		if produced.Type == "combat_tank_h" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
+	Trigger.OnProduction(CHeavyFactory1, function (producer, produced)
+		if produced.Type == "combat_tank_h" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
+	Trigger.OnProduction(CStarport, function (producer, produced)
+		if produced.Type == "combat_tank_h.starport" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
+	Trigger.OnProduction(AHeavyFactory1, function (producer, produced)
+		if produced.Type == "combat_tank_a" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
+	Trigger.OnProduction(AHeavyFactory2, function (producer, produced)
+		if produced.Type == "combat_tank_a" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
+	Trigger.OnProduction(AStarport, function (producer, produced)
+		if produced.Type == "combat_tank_a.starport" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
```
```suggestion
	local crusherTypes = { "combat_tank_a", "combat_tank_a.starport", "combat_tank_h", "combat_tank_h.starport" }
	local crusherFactories = { AStarport, AHeavyFactory1, AHeavyFactory2, CHeavyFactory1, CHeavyFactory2, CStarport }

Utils.Do(crusherFactories, function(factory)
	Trigger.OnProduction(factory, function(producer, produced)
		if not producer.Owner.IsBot then
			return
		end

		local crusher = Utils.Any(crusherTypes, function(ct) return produced.Type == ct end)

		if crusher then
			AICrushLogic(produced, producer.Owner)
		end
	end)
end)
```
I figure this should make things much easier to read, edit, and apply to other missions.

Porenutak replied:
Probably It makes even more sense to move this as function into campaign.lua ?

JovialFeline commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -277,3 +282,43 @@ ProduceUnits = function(player, factory, delay, toBuild, attackSize, attackThres
+		Trigger.AfterDelay(55, function ()
+			AICrushLogic(unit, bot)
+		end)
+		IdleHunt(unit)
```
```suggestion
		unit.Hunt()
```

JovialFeline commented on mods/d2k/maps/harkonnen-04/harkonnen04-AI.lua:
```diff
@@ -42,4 +42,11 @@ ActivateAI = function()
+		if produced.Type == "combat_tank_a" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
```
```suggestion
	ActivateCrusherOnProductions( { "combat_tank_a" }, { AHeavyFactory } )
```
Seems like a duplicate, and the undefined crusherFactories crashes on launch.

JovialFeline commented on mods/d2k/maps/harkonnen-06b/harkonnen06b-AI.lua:
```diff
@@ -33,18 +57,33 @@ ActivateAI = function()
+		ProduceUnits(OrdosSmall, OBarracks2, delay, infantryToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
+		ProduceUnits(OrdosSmall, OLightFactory2, delay, vehilcesToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
+	end)
+	ActivateCrusherOnProductions({"combat_tank_o"}, {OHeavyFactory})
```
```suggestion
	local crusherTypes = { "combat_tank_o", "combat_tank_o.starport" }
	local crusherFactories = { OHeavyFactory, OStarport }
	ActivateCrusherOnProductions(crusherTypes, crusherFactories)
```

JovialFeline commented on mods/d2k/maps/harkonnen-06a/harkonnen06a-AI.lua:
```diff
@@ -33,18 +54,59 @@ ActivateAI = function()
+			if not producer.Owner.IsBot then
+				return
+			end
+
+			local crusher = Utils.Any(crusherTypes, function(ct) return produced.Type == ct end)
+
+			if crusher then
+				AICrushLogic(produced, producer.Owner)
+			end
+		end)
+	end)
+
+	Trigger.OnProduction(OHeavyFactory1, function (producer, produced)
+		if produced.Type == "combat_tank_o" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
+	Trigger.OnProduction(OHeavyFactory2, function (producer, produced)
+		if produced.Type == "combat_tank_o" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
+	Trigger.OnProduction(OStarport, function (producer, produced)
+		if produced.Type == "combat_tank_o.starport" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
```
```suggestion
	ActivateCrusherOnProductions(crusherTypes, crusherFactories)
```

JovialFeline commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -277,3 +282,62 @@ ProduceUnits = function(player, factory, delay, toBuild, attackSize, attackThres
+	if unit.IsDead then
+		return
+	end
+	if Utils.RandomInteger(1,101) <= CrushChance[Difficulty] then
```
Unless I am extra tired, that may need inverting. On a roll of 29 and a chance of 10/Easy, a crush is attempted. If the chance is "increased" to 30/Normal or 50/Hard, that same crush is now skipped.

This would also be an excellent spot to add an owner check, in case the vehicle has been turned by a Deviator.

JovialFeline commented on mods/d2k/maps/ordos-05/ordos05-AI.lua:
```diff
@@ -37,20 +63,30 @@ InitAIUnits = function(house)
-
-	ProduceUnits(AtreidesSmall2, ABarracks3, delay, infantryToBuild, AttackGroupSize[Difficulty], attackTresholdSize)
-
+	ActivateCrusherOnProductions({"combat_tank_a","combat_tank_a.starport"}, {AStarport, AHeavyFactory})
```
Should be harmless, but the Starport is not used by the bots here or in the original. However, that did lead me to another thought.

A bonkers player could crash the game by ignoring the Starport base at first, destroying/selling the factory, and then capturing the Starport to trigger this function. It is doable, though I would not have thought to try it on this mission before, and it requires some pretty specific map knowledge.
<img width="1920" height="1080" alt="d2k-2025-07-26T042316818Z_indexed" src="https://github.com/user-attachments/assets/ccf8ad26-26c1-498f-8069-09fad098b7d7" />

JovialFeline reviewed (changes requested):
I had _some_ energy for once, so I went through a few missions and gave the code a glance. Did not get to most of the Harkonnen maps, and I think I'd rather redo the set with #21930's balance changes since those seem pretty substantial. I do like the feel of the crushing behavior as-is with the chance check fixed. :thumbsup:

----

`atreides-05`: Done on Normal. Didn't quite get Hard, though I'm sure it's doable with some Troopers dug into the right spots. The contraband hardly has time to unload before the first tank wave hits the Starport. This is still a noticeable improvement over the old version.

`harkonnen-04`: Done on Hard. Didn't have any big changes but I ran through as a bug check, considering the crash I noted in a comment. The whims of Shai-Halud have an unusual impact because it lingers near the only corridor in and out of the start base, but whaddya gonna do? I took some joy in enemy tanks being eaten while my harvesters were parked.

`harkonnen-05`: Done on Hard. Took some false starts to settle on a plan. Quite a bloodbath at the Outpost, but it held at the expense of many infantry.

`ordos-05`: Seems fine on Normal. Haven't nabbed Hard yet.

`ordos-06a`: Uncompleted. I'm unsure how to gracefully handle Siege Tanks blowing through my Troopers, and the terrain makes it rough to replace any losses in a timely manner. Abandoning the area would be an easy decision for me if the port was not a primary objective.
  - I imagine any Starport changes in the future will have a notable impact, and that `ordos-06b` will be more forgiving because its port is closer to the player's base.
  - For what it's worth, the '98 version gave me trouble as well.

JovialFeline commented on mods/d2k/maps/atreides-04/atreides04-AI.lua:
```diff
@@ -63,4 +63,6 @@ ActivateAI = function()
 	ProduceUnits(Harkonnen, HarkonnenBarracks, delay, infantryToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
 	ProduceUnits(Harkonnen, HarkonnenHeavyFact, delay, tanksToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
+
+	ActivateCrusherOnProductions({"combat_tank_h"}, {HarkonnenHeavyFact})
```
```suggestion
	ActivateCrusherOnProductions({ "combat_tank_h" }, { HarkonnenHeavyFact} )
```

JovialFeline commented on mods/d2k/maps/atreides-05/atreides05-AI.lua:
```diff
@@ -66,12 +88,21 @@ ActivateAI = function()
+		ProduceUnits(Harkonnen, HarkonnenLightFactory, delay, vehilcesToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
+		ProduceUnits(Harkonnen, HarkonnenHeavyFactory, delay, tanksToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
+	end)
+	ActivateCrusherOnProductions({"combat_tank_h"}, {HarkonnenHeavyFactory})
```
```suggestion
	ActivateCrusherOnProductions({ "combat_tank_h" }, { HarkonnenHeavyFactory })
```

JovialFeline commented on mods/d2k/maps/harkonnen-05/harkonnen05-AI.lua:
```diff
@@ -28,22 +49,33 @@ OrdosTankType = { "combat_tank_o" }
-	ProduceUnits(OrdosSmall, OBarracks3, delay, infantryToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
+	Trigger.AfterDelay(InitialProductionDelay["OrdosSmall"][Difficulty], function()
+		ProduceUnits(OrdosSmall, OBarracks3, delay, infantryToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
 	ProduceUnits(OrdosSmall, OLightFactory2, delay, vehilcesToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
```
```suggestion
		ProduceUnits(OrdosSmall, OBarracks3, delay, infantryToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
		ProduceUnits(OrdosSmall, OLightFactory2, delay, vehilcesToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
```

JovialFeline commented on mods/d2k/maps/harkonnen-06a/harkonnen06a-AI.lua:
```diff
@@ -33,18 +54,30 @@ ActivateAI = function()
+	end)
+	Trigger.AfterDelay(InitialProductionDelay["OrdosSmall"][Difficulty], function()
 	ProduceUnits(OrdosSmall, OBarracks2, delay, infantryToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
 	ProduceUnits(OrdosSmall, OLightFactory2, delay, vehilcesToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
```
```suggestion
		ProduceUnits(OrdosSmall, OBarracks2, delay, infantryToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
		ProduceUnits(OrdosSmall, OLightFactory2, delay, vehilcesToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
```

JovialFeline commented on mods/d2k/maps/harkonnen-06b/harkonnen06b-AI.lua:
```diff
@@ -24,7 +46,9 @@ AttackDelays =
-OrdosStarportTypes = { "trike.starport", "trike.starport", "quad.starport", "combat_tank_o.starport", "combat_tank_o.starport", "siege_tank.starport", "missile_tank.starport" }
+OrdosStarportTypes = {
+	EarlyGame =  {"trike.starport", "trike.starport", "quad.starport", "combat_tank_o.starport", "combat_tank_o.starport", "siege_tank.starport"},
+	LateGame =  {"trike.starport", "quad.starport", "combat_tank_o.starport", "siege_tank.starport", "missile_tank.starport", "missile_tank.starport"} }
```
```suggestion
	EarlyGame =  { "trike.starport", "trike.starport", "quad.starport", "combat_tank_o.starport", "combat_tank_o.starport", "siege_tank.starport" },
	LateGame =  { "trike.starport", "quad.starport", "combat_tank_o.starport", "siege_tank.starport", "missile_tank.starport", "missile_tank.starport" } }
```

JovialFeline commented on mods/d2k/maps/harkonnen-07/harkonnen07-AI.lua:
```diff
@@ -6,31 +6,54 @@
+AtreidesMainTankTypes =
+{
+	EarlyGame = {"combat_tank_a", "combat_tank_a", "siege_tank"},
+	LateGame = {"combat_tank_a", "siege_tank", "missile_tank", "sonic_tank" }
```
```suggestion
	EarlyGame = { "combat_tank_a", "combat_tank_a", "siege_tank" },
	LateGame = { "combat_tank_a", "siege_tank", "missile_tank", "sonic_tank" }
```

JovialFeline replied:
There's a decent number of these throughout, so I won't comment on all of them.

JovialFeline commented on mods/d2k/maps/harkonnen-09a/harkonnen09a-AI.lua:
```diff
@@ -7,31 +7,78 @@
-
+AtreidesStarportTypes = {
+	EarlyGame ={ "trike.starport", "trike.starport", "trike.starport", "quad.starport","quad.starport", "combat_tank_a.starport"},
+	LateGame ={ "trike.starport", "trike.starport", "quad.starport", "combat_tank_a.starport", "combat_tank_a.starport", "siege_tank.starport", "missile_tank.starport"}
```
```suggestion
	EarlyGame = { "trike.starport", "trike.starport", "trike.starport", "quad.starport", "quad.starport", "combat_tank_a.starport" },
	LateGame = { "trike.starport", "trike.starport", "quad.starport", "combat_tank_a.starport", "combat_tank_a.starport", "siege_tank.starport", "missile_tank.starport" }
```

JovialFeline commented on mods/d2k/maps/harkonnen-06b/harkonnen06b-AI.lua:
```diff
@@ -24,7 +46,9 @@ AttackDelays =
+OrdosStarportTypes = {
+	EarlyGame =  {"trike.starport", "trike.starport", "quad.starport", "combat_tank_o.starport", "combat_tank_o.starport", "siege_tank.starport"},
+	LateGame =  {"trike.starport", "quad.starport", "combat_tank_o.starport", "siege_tank.starport", "missile_tank.starport", "missile_tank.starport"} }
 
```
```suggestion
OrdosStarportTypes =
{
	EarlyGame =  { "trike.starport", "trike.starport", "quad.starport", "combat_tank_o.starport", "combat_tank_o.starport", "siege_tank.starport" },
	LateGame =  { "trike.starport", "quad.starport", "combat_tank_o.starport", "siege_tank.starport", "missile_tank.starport", "missile_tank.starport" }
}
```

JovialFeline commented on mods/d2k/maps/harkonnen-07/harkonnen07-AI.lua:
```diff
@@ -45,23 +68,38 @@ ActivateAI = function()
+			return { Utils.Random(AtreidesMainTankTypes["EarlyGame"]) }
+		else
+			return { Utils.Random(AtreidesMainTankTypes["LateGame"]) }
+		end end
```
```suggestion
		end
	end
```

JovialFeline commented on mods/d2k/maps/harkonnen-08/harkonnen08-AI.lua:
```diff
@@ -24,12 +52,23 @@ AttackDelays =
-AtreidesStarportTypes = { "trike.starport", "trike.starport", "quad.starport", "combat_tank_a.starport", "combat_tank_a.starport", "siege_tank.starport", "missile_tank.starport" }
+AtreidesTankTypes = {
+	EarlyGame ={ "combat_tank_a" },
+	LateGame ={ "combat_tank_a", "combat_tank_a", "siege_tank" }}
```
```suggestion
AtreidesTankTypes = {
	EarlyGame = { "combat_tank_a" },
	LateGame = { "combat_tank_a", "combat_tank_a", "siege_tank" }
}
```

JovialFeline commented on mods/d2k/maps/harkonnen-09b/harkonnen09b-AI.lua:
```diff
@@ -7,30 +7,75 @@
+EnemyVehicleTypes = { "trike", "trike", "quad" }
+AtreidesMainTankTypes = {
+	EarlyGame ={"combat_tank_a", "combat_tank_a", "siege_tank", "missile_tank"},
+	LateGame = {"combat_tank_a", "combat_tank_a", "siege_tank", "missile_tank", "sonic_tank"}
```
```suggestion
	EarlyGame = { "combat_tank_a", "combat_tank_a", "siege_tank", "missile_tank" },
	LateGame = { "combat_tank_a", "combat_tank_a", "siege_tank", "missile_tank", "sonic_tank" }
```

JovialFeline commented on mods/d2k/maps/harkonnen-08/harkonnen08-AI.lua:
```diff
@@ -24,12 +52,23 @@ AttackDelays =
+	LateGame ={ "combat_tank_a", "combat_tank_a", "siege_tank" }}
+AtreidesStarportTypes = {
+	EarlyGame ={"trike.starport", "trike.starport", "quad.starport", "combat_tank_a.starport", "combat_tank_a.starport"},
+	LateGame ={"trike.starport", "quad.starport", "combat_tank_a.starport", "combat_tank_a.starport", "siege_tank.starport", "missile_tank.starport"} }
```
```suggestion
	EarlyGame = { "trike.starport", "trike.starport", "quad.starport", "combat_tank_a.starport", "combat_tank_a.starport" },
	LateGame = { "trike.starport", "quad.starport", "combat_tank_a.starport", "combat_tank_a.starport", "siege_tank.starport", "missile_tank.starport" }
}
```

JovialFeline commented on mods/d2k/maps/harkonnen-09b/harkonnen09b-AI.lua:
```diff
@@ -7,30 +7,75 @@
-
+AtreidesStarportTypes = {
+	EarlyGame ={ "trike.starport", "trike.starport", "trike.starport", "quad.starport","quad.starport", "combat_tank_a.starport"},
+	LateGame ={ "trike.starport", "trike.starport", "quad.starport", "combat_tank_a.starport", "combat_tank_a.starport", "siege_tank.starport", "missile_tank.starport"}
```
```suggestion
	EarlyGame = { "trike.starport", "trike.starport", "trike.starport", "quad.starport", "quad.starport", "combat_tank_a.starport" },
	LateGame = { "trike.starport", "trike.starport", "quad.starport", "combat_tank_a.starport", "combat_tank_a.starport", "siege_tank.starport", "missile_tank.starport" }
```

JovialFeline commented on mods/d2k/maps/harkonnen-09b/harkonnen09b-AI.lua:
```diff
@@ -42,32 +87,74 @@ ActivateAI = function()
-	local tanksToBuildAtreidesMain = function() return { Utils.Random(AtreidesMainTankTypes) } end
+	local tanksToBuildAtreidesMain = function()
+		if EarlyGameStage >= DateTime.GameTime then
+		return { Utils.Random(AtreidesMainTankTypes["EarlyGame"]) }
```
```suggestion
			return { Utils.Random(AtreidesMainTankTypes["EarlyGame"]) }
```

JovialFeline commented on mods/d2k/maps/harkonnen-09b/harkonnen09b-AI.lua:
```diff
@@ -42,32 +87,74 @@ ActivateAI = function()
-	local unitsToBuyCorrino = function() return { Utils.Random(CorrinoStarportTypes) } end
+	local tanksToBuildCorrino = function()
+		if EarlyGameStage >= DateTime.GameTime then
+		return { Utils.Random(CorrinoTankTypes["EarlyGame"]) }
```
```suggestion
			return { Utils.Random(CorrinoTankTypes["EarlyGame"]) }
```

JovialFeline commented on mods/d2k/maps/harkonnen-09a/harkonnen09a-AI.lua:
```diff
@@ -7,31 +7,78 @@
 
-AtreidesMainTankTypes = { "combat_tank_a", "combat_tank_a", "siege_tank", "missile_tank", "sonic_tank" }
+AtreidesMainTankTypes = {
+	EarlyGame ={"combat_tank_a", "combat_tank_a", "siege_tank", "missile_tank"},
```
```suggestion
	EarlyGame = { "combat_tank_a", "combat_tank_a", "siege_tank", "missile_tank" },
```

JovialFeline commented on mods/d2k/maps/harkonnen-09a/harkonnen09a-AI.lua:
```diff
@@ -46,35 +93,74 @@ ActivateAI = function()
-	ProduceUnits(AtreidesSmall1, AHeavyFactory2, delay, tanksToBuildAtreidesSmall, AttackGroupSize[Difficulty], attackThresholdSize)
+	local tanksToBuildCorrinoMain = function()
+		if EarlyGameStage >= DateTime.GameTime then
+		return { Utils.Random(CorrinoMainTankTypes["EarlyGame"]) }
```
```suggestion
			return { Utils.Random(CorrinoMainTankTypes["EarlyGame"]) }
```

JovialFeline commented on mods/d2k/maps/harkonnen-09a/harkonnen09a-AI.lua:
```diff
@@ -46,35 +93,74 @@ ActivateAI = function()
-	local tanksToBuildAtreidesMain = function() return { Utils.Random(AtreidesMainTankTypes) } end
+	local tanksToBuildAtreidesMain = function()
+		if EarlyGameStage >= DateTime.GameTime then
+		return { Utils.Random(AtreidesMainTankTypes["EarlyGame"]) }
```
```suggestion
			return { Utils.Random(AtreidesMainTankTypes["EarlyGame"]) }
```

JovialFeline reviewed (commented):
I seem to be bad at D2k, which isn't the worst thing for playtesting this. I did eventually lose my appetite and stop making attempts on Hard though.

A side effect of [the Starport overhaul](https://github.com/OpenRA/OpenRA/pull/21424) may be making later Harkonnen missions a bit easier than intended. Bots seem unable to Build from a port since those changes, which I did not realize until after this batch of playtests.

----

### Harkonnen 04

Beat on Normal.
- Holding the north-northwest chokepoint with covered infantry helps with harvester harassment.

Three Hard attempts made.
- Last attempt was closest, until the eighth Fremen wave (composed entirely of tanks) cut off the spice and destroyed my remaining army around 9:04.
- May benefit from an attack delay. The first wave will coincide with a Fremen tank attack and hit before a barracks and refinery are up.

Compared to this, the bots in D2k '98 felt oddly reluctant to aggress. In that run, the south end of the base was only hit with one infantry wave.

### Harkonnen 05

Done on Normal.
- Not too rough. Just need to be careful about quad/tank attacks on harvesters and keep the outpost guarded.

Done on Hard. Okay time at 26:07.
- Swapped some of the outpost infantry from my Normal run for a turret. That seemed to do a better job of holding until the southern base could be properly attacked.

### Harkonnen 06a

Done on Normal. Took my time at 41m.

### Harkonnen 06b

Done on Normal at a decent pace, 19:44.
- Helps to concentrate refineries on the southern end; attacks at the east corridor are constant. Pushed through to focus on the Starport, with infantry up front to distract any Deviators.

Attempted on Hard.
- Start is hot. Expect to see tanks once the refinery is up.

This map doesn't give tank reinforcements at the start like the A side, but that's how the original works too.

### Harkonnen 07

Completed on Easy. Have not yet figured out a good way to hold the start on Normal.

### Harkonnen 08

Completed on Easy.
- Swift capture of the Mercenary base is tricky but it relieves a lot of pressure.
- Deviators and their post-death gas cloud give the Mercenaries more trouble than one might think.

### Harkonnen 09ab

Practically untested. Failed on Easy, mostly due to artillery. Will try again.

### Ordos 05

Finished Normal after one restart. Did not expand. 24m.
- Lots of infantry sacrificed to hold northwest strip of rock.

Attempted on Hard with rough results.

### Ordos 06a

Untested. Not in a hurry to repeat the experience from last time.

### Atreides 04

Finished on Hard. 16:32.
- Running a mass of starting units through the enemy base to the Sietch seems nearly required.
- Fremen can sometimes follow the trail of enemy units back to a Barracks or Heavy Factory. Taking advantage of this to knock out production will be less effective when bot rebuilds are introduced.

### Atreides 05

Finished on Normal. Hard remains a bit outside my reach, though the loss was more of a slow grind than a blowout.

---

Other tangentially related stuff in my notes:

**Atreides 04**
Produced Fremen should probably use a rally point but this would have little effect on difficulty. It may be most noticeable on Hard where a lone Fremen can stall for the 100% integrity objective. Trikes in particular like to take advantage of a perimeter gap.

**Harkonnen 07**
The old script was not made with rebuilds in mind. Whenever bot rebuilding becomes a thing, the enemy Carryall waves should restart when the High Tech Factory is rebuilt. The condition should change to check the status of the _current_ factory to avoid duplicate reinforcements.

**Harkonnen 06 - 09**
The neutral-to-hostile checks could be more picky about damage sources. It's possible that an "attack" from harvester/building shrapnel sets off the trigger. An Outpost in Ordos 04 used to have [a similar issue, though with lower stakes](https://github.com/OpenRA/OpenRA/pull/21245).

Nitpicks follow.

Porenutak commented:
Nitpicks fixed. For a future It will be good idea to use force style rules same as in c# code.

There is lots of thinks to address here, so for now Im marking this as Draft until its ready for review again.

> Compared to this, the bots in D2k '98 felt oddly reluctant to aggress. In that run, the south end of the base was only hit with one infantry wave.

This is because of flaws in d2k script design. To address this I would need to delete everything in campaign.lua and start from the scratch. That Im not willing to do.

Porenutak commented:
This PR is ready for review again.

As I play missions more and compared it to OG. I realize that main problem is that AI can do only one thing: IdleHunt. Resolving the first attack to be devastating but late game more less stale. And we want this to be exact opposite. 
This PR now rework most of the  LUA scripts, implementing lots AI behaviour from OG game (or inspired by):

-	Each faction implements `InitialAttackDelay `and `TimeBetweenAttacks `from OG game
-	If AI want attack but `TimeBetweenAttacks `is still in place, it create patrol squad and go patrolling instead. Production is on hold until `TimeBetweenAttacks `expire. If patrol squad is under attack it defend self, but never rush base or go on non-stop IdleHunt. When patrol ends, squad return to the `IdlingUnits `pool
- Idling units are moving around the base searching for enemy, when killed, AI will check the area. (This partly mimics GuardBase from OG)
- Rename `AttackDelays `to `ProductionDelays`,  because thats what it does.
- If AI is under pressure, it can activated Emergecy Mode removing ProductionDelays between queuing or add another mission specific bonus
- Small base can ask Main base for help. If Main base has enough Idling units.
- AI maintain desired number of harvesters
- Airstrike rework: AI can use airstrike vs buildings or use it as close combat support during emergencyMode 
- Saboteur can demolish vehicles
- add correct starting cash for each bot.

As early games are now more quieter I decide to add some small nitpicks that wasnt in OG missions.
- crates
- destroyable debris that gives you a crate.
- husks on places where it make sense.
There  are not in every map and serve as something as reward for  exploring or easter eggs. This doenst effect game play in any way.

But there is one bigger change:
IN ORA D2k engineers can capture husk. I feel we need to introduce this feature somehow, otherwise only 2 ppl playing MP will know about it :)
At least one mission for each faction should introduce this feature in subtle way. I add Mission message informing player about this in harkonnen9b and atreides5. In these maps you start without MCV but with repair pad. 

Other changes:
- mission harkonnen7 and harkonnen9a got more reinforcements, because even with all changes above the first attack was almost impossible to repel. 
- mission4 on hard: 100 integrity for sietch is kinda insane. And it wasnt even in OG mission. I lower the value to 80%
- smugglers Relations change: you need to do 20% damage to the actor for relations change  to hostile. 

Played all missions on Hard difficulty and fastest speed and all missions are winnable (unlike before this PR), but I of course know every inside so maybe difficulty can be tweaked even more.

As much as I try this to by bug free, there is so much rework and so much missions (with so much options) that it is impossible to guarantee everything will work.
I didnt remove my debug messages yet, I will after this get approval. There is also DebugInfo() function if you wanna inside during gameplay.

Porenutak commented:
Forgot to mention: 
AI can rebuild base. On hard difficulty it can also rebuild defenses

PunkPun commented on mods/d2k/fluent/lua.ftl:
```diff
@@ -76,6 +76,7 @@ smuggler-leader = Smuggler Leader.
 were-warned-will-pay = You were warned. Now you will pay.
 destroy-smugglers-mercenaries = Destroy the Smugglers and their mercenaries.
 capture-harkonnen-barracks-release-hostages = Capture the Harkonnen barracks to release the hostages.
+use-engineer-hint = Use engineers to reclam destroyed vehicles
```
```suggestion
use-engineer-hint = Use engineers to reclam destroyed vehicles.
```

Please add punctuations.

PunkPun commented on mods/d2k/fluent/lua.ftl:
```diff
@@ -150,3 +151,11 @@ additional-reinforcements-arrive-in = Additional reinforcements will arrive in {
+
+## harkonnen-9b
+husk-reclaimed-harkonnen = We reclaimed some Atreides wreckage from battle
+engineers-hint = Our engineers maybe find then useful
```
Please place these in correct places and use punctuation

PunkPun commented on mods/d2k/maps/atreides-02a/atreides02a-AI.lua:
```diff
@@ -25,10 +25,13 @@ HarkonnenInfantryTypes = { "light_inf" }
-	local delay = function() return Utils.RandomInteger(AttackDelays[Difficulty][1], AttackDelays[Difficulty][2] + 1) end
+	local delay = function() return Utils.RandomInteger(ProductionDelays[Difficulty][1], ProductionDelays[Difficulty][2] + 1) end
 	local toBuild = function() return HarkonnenInfantryTypes end
-	local attackThresholdSize = AttackGroupSize[Difficulty] * 2.5
```
unnecesary change

PunkPun commented on mods/d2k/maps/atreides-02a/atreides02a-AI.lua:
```diff
@@ -25,10 +25,13 @@ HarkonnenInfantryTypes = { "light_inf" }
 	IdlingUnits[Harkonnen] = { }
+	Defending[Harkonnen] = { }
+	AttackDelay[Harkonnen] = 6000 * DifficultyModifier[Difficulty]
+	TimeBetweenAttacks[Harkonnen] = 5000 * DifficultyModifier[Difficulty]
```
this will be a pain to maintain. Setting magic values for a script that doesn't even exist here. Rather we should call the functions here

Porenutak replied:
I dont get this. AttackDelay and TimeBetweenAttacks are used in ProduceUnits function. Numbers are taken from D2keditor. What do you have in mind?

PunkPun commented on mods/ra/maps/allies-04/allies04-AI.lua:
```diff
@@ -18,7 +18,7 @@ AttackGroupSizes =
 }
 
-AttackDelays =
+AttackDelays  =
```
why touch RA? And with en erroneous fix even

PunkPun commented on mods/d2k/sequences/vehicles.yaml:
```diff
@@ -353,6 +353,18 @@ combat_tank_a.husk:
+		ZOffset: -512
+	tankdebris04:
+		Filename: missiletankhusk/missiletankhuskdebris-0003.png
+		ZOffset: -512
```
these look quite weird, what's the idea here?

Porenutak replied:
Sorry, forgot to delete

PunkPun commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +355,305 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+        end
+    end
+    return false
+end
```
instead add this to UtilsGlobal?

```csharp
[Desc("Returns true if the collection contains the specified value.")]
public bool Contains(
	[ScriptEmmyTypeOverride("T[]", "T")] LuaValue[] collection,
	[ScriptEmmyTypeOverride("T", "T")] LuaValue value)
{
	return collection.Any(c => c.Equals(value));
}
```

Porenutak replied:
It would be even better if Trigger.OnAnyProduction returns also buildings. Or we have something like Trigger.OnBuildingPlaced.
Then whole FindNewBuilding function woundnt be needed.
But thats out of scope for this PR.

Porenutak replied:
> ```cs
> 	[ScriptEmmyTypeOverride("T", "T")] LuaValue value)
> ```

For whatever reason the second parameter is never recognized as LuaValue. What can be a problem?

PunkPun replied:
#22164 was merged

JovialFeline commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +355,305 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+					if #productionTypes > 0 then
+						Media.Debug("production type"..productionTypes[1])
+						local productionToBuild = function() return { Utils.Random(productionTypes) } end
+						ProduceUnits(player, building, delay, productionToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
```
I suspect it would be simpler (at least in the long run) to pass along a function to buildingsToRebuild that resumes production through this type and calls DefendAndRepairBase rather than passing the delay and attackThresholdSize.

Either way, it looks like new factories won't use the crusher logic.

JovialFeline commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +355,305 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+	if unit.IsDead then return end
+	local targetCell = Utils.Random(DefencePerimeter[owner])
+	unit.AttackMove(targetCell, 2)
+	unit.Wait(Utils.RandomInteger(400, 800))
```
I've found that Wait acts more like a Sleep command. Armed infantry or non-turreted vehicles will be unable to attack while Waiting. This also applies to Patrol's `wait` ticks.

Porenutak replied:
I know. Adding Trigger.OnDamage can resolve this problem. But isnt it too performace heavy for unit that just idling in a base?
Wait order is canceled when any unit is killed or other event happend

PunkPun replied:
Trigger.AfterDelay ?

JovialFeline commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +355,305 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+	Trigger.OnAllKilledOrCaptured(baseChecked, function()
+		if Utils.Any(base, function(building) return not building.IsDead end)
+		then
+			IdleHuntOnBaseDestroyed(player, base)
```
```suggestion
	local checkedBase = Utils.Where(base, function(building) return not building.IsDead and building.Owner == player end)
	Trigger.OnAllKilledOrCaptured(checkedBase, function()
		if Utils.Any(checkedBase, function(building) return not building.IsDead and building.Owner == player end)
		then
			IdleHuntOnBaseDestroyed(player, checkedBase)
```
Unless I'm misunderstanding the intent.

JovialFeline commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +355,305 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+				and not a.IsDead
+				and not a.Owner.IsAlliedWith(unit.Owner)
+				and a.Owner.InternalName ~= "Neutral"
+				and a.Owner.InternalName ~= "Creep"
```
```suggestion
				and a.Owner.InternalName ~= "Creeps"
```

JovialFeline commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +355,305 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+		a.Type == "trooper" or
+		a.Type == "engineer" and
+		Map.TerrainType(a.Location) ~= "Rough"
 	end)
```
```suggestion
	local targets = Map.ActorsInCircle(unit.CenterPosition, WDist.FromCells(5), function (a)
		return
		a.Owner.IsAlliedWith(bot) == false and
		a.Type == "light_inf" or
		a.Type == "trooper" or
		a.Type == "engineer" and
		Map.TerrainType(a.Location) ~= "Rough"
	end)
```

JovialFeline reviewed (commented):
That's a lot of changes to dig into, so this is just some things I have at a glance.

PunkPun commented:
needs a rebase

pchote commented:
I pushed a rebase + fixed typos in the commit message.

pchote commented on mods/d2k/maps/atreides-04/atreides04-AI.lua:
```diff
@@ -33,15 +33,22 @@ SendAttack = function(owner, size)
+		Trigger.OnAllRemovedFromWorld(units, function()
+			Attacking[owner] = false
+			HoldProduction[owner] = false
+			Media.Debug("attack stoped")
```
Remove this?

pchote commented on mods/d2k/maps/atreides-03a/atreides03a-AI.lua:
```diff
@@ -49,4 +53,20 @@ ActivateAI = function()
+	}
+	Trigger.OnBuildingPlaced(Ordos, function(p, building)
+		table.insert(OrdosBase, building)
+		DefendAndRepairBase(Ordos, {building},0.75, AttackGroupSize[Difficulty] )
```
Missing space between `{building},0.75`

pchote commented on mods/d2k/maps/atreides-03b/atreides03b-AI.lua:
```diff
@@ -34,12 +39,34 @@ ActivateAI = function()
+	}
+	Trigger.OnBuildingPlaced(Ordos, function(p, building)
+		table.insert(OrdosBase, building)
+		DefendAndRepairBase(Ordos, {building},0.75, AttackGroupSize[Difficulty] )
```
Missing space between `{building},0.75`

pchote commented on mods/d2k/maps/atreides-03b/rules.yaml:
```diff
@@ -1,6 +1,33 @@
+			barracks: 1
+			refinery: 1
+			outpost: 1
+
```
Stray newline

pchote commented on mods/d2k/maps/atreides-04/atreides04-AI.lua:
```diff
@@ -52,15 +59,32 @@ InitAIUnits = function()
+	}
+	Trigger.OnBuildingPlaced(Harkonnen, function(p, building)
+		table.insert(HarkonnenBase, building)
+		DefendAndRepairBase(Harkonnen, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/atreides-05/atreides05-AI.lua:
```diff
@@ -25,6 +40,7 @@ HarkonnenInfantryTypes = { "light_inf", "light_inf", "trooper", "trooper", "troo
 HarkonnenVehicleTypes = { "trike", "trike", "trike", "quad", "quad" }
 HarkonnenTankType = { "combat_tank_h" }
 
+
```
Stray newline

pchote commented on mods/d2k/maps/atreides-05/rules.yaml:
```diff
@@ -1,6 +1,64 @@
+		Condition: base-rebuilder
+	ExternalCondition@AIDefense:
+		Condition: defense-rebuilder
+	BaseBuilderBotModule@campaing:
```
typo: `campaing`

pchote commented on mods/d2k/maps/harkonnen-03b/rules.yaml:
```diff
@@ -1,6 +1,34 @@
 		DefaultCash: 5000
+	ExternalCondition@AI:
+		Condition: base-rebuilder
+	BaseBuilderBotModule@campaing:
```
typo: `campaing`

pchote commented on mods/d2k/maps/harkonnen-05/rules.yaml:
```diff
@@ -1,6 +1,68 @@
+		Condition: base-rebuilder
+	ExternalCondition@AIDefense:
+		Condition: defense-rebuilder
+	BaseBuilderBotModule@campaing:
```
typo: `campaing`

pchote commented on mods/d2k/maps/harkonnen-03a/rules.yaml:
```diff
@@ -1,6 +1,34 @@
 		DefaultCash: 5000
+	ExternalCondition@AI:
+		Condition: base-rebuilder
+	BaseBuilderBotModule@campaing:
```
typo: `campaing`

pchote commented on mods/d2k/maps/atreides-05/atreides05-AI.lua:
```diff
@@ -63,15 +86,49 @@ end
+	}
+	Trigger.OnBuildingPlaced(Harkonnen, function(p, building)
+		table.insert(HarkonnenBase, building)
+		DefendAndRepairBase(Harkonnen, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-03a/harkonnen03a-AI.lua:
```diff
@@ -42,4 +46,18 @@ ActivateAI = function()
+	}
+	Trigger.OnBuildingPlaced(Atreides, function(p, building)
+		table.insert(OrdosMainBase, building)
+		DefendAndRepairBase(Atreides, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-03b/harkonnen03b-AI.lua:
```diff
@@ -47,4 +50,19 @@ ActivateAI = function()
+
+	Trigger.OnBuildingPlaced(Atreides, function(p, building)
+		table.insert(OrdosMainBase, building)
+		DefendAndRepairBase(Atreides, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-04/harkonnen04-AI.lua:
```diff
@@ -6,40 +6,88 @@
+	}
+	Trigger.OnBuildingPlaced(Atreides, function(p, building)
+		table.insert(AtreidesBase, building)
+		DefendAndRepairBase(Atreides, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-05/harkonnen05-AI.lua:
```diff
@@ -7,43 +7,109 @@
+	}
+	Trigger.OnBuildingPlaced(OrdosMain, function(p, building)
+		table.insert(OrdosMainBase, building)
+		DefendAndRepairBase(OrdosMain, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-06a/harkonnen06a-AI.lua:
```diff
@@ -27,24 +48,76 @@ OrdosTankTypes = { "combat_tank_o", "combat_tank_o", "siege_tank" }
-	ProduceUnits(OrdosSmall, OLightFactory2, delay, vehilcesToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
+	Trigger.OnBuildingPlaced(OrdosSmall, function(p, building)
+		table.insert(OrdosSmallBase, building)
+		DefendAndRepairBase(OrdosSmall, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-06b/harkonnen06b-AI.lua:
```diff
@@ -24,27 +39,93 @@ AttackDelays =
+	}
+	Trigger.OnBuildingPlaced(OrdosMain, function(p, building)
+		table.insert(OrdosMainBase, building)
+		DefendAndRepairBase(OrdosMain, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-06b/harkonnen06b-AI.lua:
```diff
@@ -24,27 +39,93 @@ AttackDelays =
-	ProduceUnits(OrdosSmall, OLightFactory2, delay, vehilcesToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
+	Trigger.OnBuildingPlaced(OrdosSmall, function(p, building)
+		table.insert(OrdosSmallBase, building)
+		DefendAndRepairBase(OrdosSmall, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-07/harkonnen07-AI.lua:
```diff
@@ -45,23 +91,70 @@ ActivateAI = function()
+	end)
+	Trigger.OnBuildingPlaced(AtreidesSmall, function(p, building)
+		table.insert(AtreidesSmallBase, building)
+		DefendAndRepairBase(AtreidesSmall, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-08/harkonnen08-AI.lua:
```diff
@@ -44,26 +112,101 @@ ActivateAI = function()
+	end)
+	Trigger.OnBuildingPlaced(AtreidesEnemy, function(p, building)
+		table.insert(AtreidesBase, building)
+		DefendAndRepairBase(AtreidesEnemy, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/ordos-03a/ordos03a-AI.lua:
```diff
@@ -47,4 +55,19 @@ ActivateAI = function()
+	}
+	Trigger.OnBuildingPlaced(Harkonnen, function(p, building)
+		table.insert(HarkonnenBase, building)
+		DefendAndRepairBase(Harkonnen, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/ordos-03b/ordos03b-AI.lua:
```diff
@@ -47,4 +60,19 @@ ActivateAI = function()
+	}
+	Trigger.OnBuildingPlaced(Harkonnen, function(p, building)
+		table.insert(HarkonnenBase, building)
+		DefendAndRepairBase(Harkonnen, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/ordos-04/ordos04-AI.lua:
```diff
@@ -55,4 +73,25 @@ ActivateAI = function()
+	}
+	Trigger.OnBuildingPlaced(Harkonnen, function(p, building)
+		table.insert(Base[Harkonnen], building)
+		DefendAndRepairBase(Harkonnen, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/ordos-05/ordos05-AI.lua:
```diff
@@ -37,26 +64,72 @@ InitAIUnits = function(house)
+	}
+	Trigger.OnBuildingPlaced(AtreidesMain, function(p, building)
+		table.insert(Base[AtreidesMainBase], building)
+		DefendAndRepairBase(AtreidesMain, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/ordos-06a/ordos06a-AI.lua:
```diff
@@ -35,20 +56,78 @@ InitAIUnits = function(house)
+	}
+	Trigger.OnBuildingPlaced(Atreides, function(p, building)
+		table.insert(Base["Atreides"], building)
+		DefendAndRepairBase(Atreides, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/ordos-06a/ordos06a-AI.lua:
```diff
@@ -35,20 +56,78 @@ InitAIUnits = function(house)
+	end)
+	Trigger.OnBuildingPlaced(Harkonnen, function(p, building)
+		table.insert(Base["Harkonnen"], building)
+		DefendAndRepairBase(Harkonnen, {building},0.75, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/atreides-05/atreides05-AI.lua:
```diff
@@ -49,7 +65,14 @@ ProduceInfantry = function()
+		delay =  Utils.RandomInteger(LateProductionDelays[Difficulty][1], LateProductionDelays[Difficulty][2] + 1)
+	end
+
+
```
stray newline

pchote commented on mods/d2k/maps/atreides-05/atreides05-AI.lua:
```diff
@@ -63,15 +86,49 @@ end
+	AttackDelay[Harkonnen] = 6000 * DifficultyModifier[Difficulty]
+	TimeBetweenAttacks[Harkonnen] = 5000 * DifficultyModifier[Difficulty]
+	HarvesterCount[Harkonnen] = 1
+	PatrolPoints[Harkonnen] = {HPatrolPoint1.Location, HPatrolPoint2.Location,HPatrolPoint3.Location}
```
missing space

pchote commented on mods/d2k/maps/atreides-05/map.yaml:
```diff
@@ -365,6 +365,41 @@ Actors:
+		Facing: 384
+		Location: 8,58
+	Trigger1: waypoint
+		Owner: Creeps
```
This should probably be Neutral?

pchote commented on mods/d2k/maps/ordos-03b/map.yaml:
```diff
@@ -254,6 +254,16 @@ Actors:
+		Facing: 384
+		Location: 30,41
+	HPatrolPoint1: waypoint
+		Owner: Creeps
```
These should probably be Neutral?

pchote commented on mods/d2k/maps/atreides-02a/map.yaml:
```diff
@@ -168,6 +168,13 @@ Actors:
+		Facing: 384
+		Location: 18,18
+	Actor25: crate
+		Owner: Creeps
```
This should probably be Neutral?

pchote commented on mods/d2k/maps/harkonnen-03a/harkonnen03a-AI.lua:
```diff
@@ -14,25 +14,29 @@ AttackGroupSize =
 	hard = { DateTime.Seconds(1), DateTime.Seconds(5) }
 }
 
+
```
stray newline

pchote commented on mods/d2k/maps/harkonnen-03b/harkonnen03b-AI.lua:
```diff
@@ -14,13 +14,14 @@ AttackGroupSize =
 	hard = { DateTime.Seconds(1), DateTime.Seconds(5) }
 }
 
+
```
stray newline

pchote commented on mods/d2k/maps/harkonnen-03b/harkonnen03b-AI.lua:
```diff
@@ -31,13 +32,15 @@ InitAIUnits = function()
 	AConyard.Produce(HarkonnenUpgrades[2])
 
-	local delay = function() return Utils.RandomInteger(AttackDelays[Difficulty][1], AttackDelays[Difficulty][2] + 1) end
+	local delay = function(plaer) return Utils.RandomInteger(ProductionDelays[Difficulty][1], ProductionDelays[Difficulty][2] + 1) end
```
typo `plaer`

pchote commented on mods/d2k/maps/harkonnen-04/harkonnen04-AI.lua:
```diff
@@ -6,40 +6,88 @@
 	hard = { DateTime.Seconds(1), DateTime.Seconds(3) }
 }
 
+
```
stray newlines

pchote commented on mods/d2k/maps/harkonnen-04/harkonnen04.lua:
```diff
@@ -105,6 +105,9 @@ FremenHunterPaths =
 	{ FremenEntry3.Location, FremenRally3.Location }
 }
 
+
```
stray newline

pchote commented on mods/d2k/maps/harkonnen-04/harkonnen04.lua:
```diff
@@ -150,25 +153,25 @@ Tick = function()
 		end
 	end
+
+	DebugInfo()
```
Remove debug code?

pchote commented on mods/d2k/maps/harkonnen-05/harkonnen05-AI.lua:
```diff
@@ -7,43 +7,109 @@
 }
 
-AttackDelays =
+
```
newlines here and 2x below

pchote commented on mods/d2k/maps/harkonnen-05/harkonnen05.lua:
```diff
@@ -282,4 +291,5 @@ WorldLoaded = function()
 	local ordosCondition = function() return Harkonnen.IsObjectiveCompleted(KillOrdos) end
 	TriggerCarryallReinforcements(Harkonnen, OrdosMain, BaseAreaTriggers[1], OrdosHunters[1], OrdosHunterPaths[2], ordosCondition)
 	TriggerCarryallReinforcements(Harkonnen, OrdosMain, BaseAreaTriggers[2], OrdosHunters[2], OrdosHunterPaths[1], ordosCondition)
+
```
stray newline

pchote commented on mods/d2k/maps/harkonnen-06a/harkonnen06a-AI.lua:
```diff
@@ -27,24 +48,76 @@ OrdosTankTypes = { "combat_tank_o", "combat_tank_o", "siege_tank" }
 ActivateAI = function()
+	Defending[OrdosMain] = {}
+	Defending[OrdosSmall] = {}
+	-- this is initial attack kdelay timing
```
typo + redundant comment?

pchote commented on mods/d2k/maps/harkonnen-06a/harkonnen06a.lua:
```diff
@@ -148,10 +155,60 @@ ChangeOwner = function(old_owner, new_owner)
+		 end
+	end)
+	end
+
```
newline

pchote commented on mods/d2k/maps/harkonnen-06a/harkonnen06a.lua:
```diff
@@ -148,10 +155,60 @@ ChangeOwner = function(old_owner, new_owner)
+	unit.Stop()
+	unit.AttackMove(intruder.Location,1)
+	Trigger.AfterDelay(100, function() FollowIntruder(unit, intruder) end)
+
```
newline

pchote commented on mods/d2k/maps/harkonnen-06a/harkonnen06a.lua:
```diff
@@ -234,6 +310,9 @@ Tick = function()
 		end
 	end
+	if IdlingUnits[OrdosMain] == nil and IdlingUnits[OrdosSmall] == nil then return end
+		UserInterface.SetMissionText("Main Base:      "..tostring(#IdlingUnits[OrdosMain]).."\n".."Small base:      "..tostring(#IdlingUnits[OrdosSmall]))
```
String needs to be extracted to fluent.

pchote commented on mods/d2k/maps/harkonnen-06b/harkonnen06b-AI.lua:
```diff
@@ -24,27 +39,93 @@ AttackDelays =
 ActivateAI = function()
+	Defending[OrdosMain] = {}
+	Defending[OrdosSmall] = {}
+	-- this is initial attack kdelay timing
```
typo/redundant.

pchote commented on mods/d2k/maps/harkonnen-06b/harkonnen06b-AI.lua:
```diff
@@ -24,27 +39,93 @@ AttackDelays =
 	IdlingUnits[OrdosMain] = Utils.Concat(Reinforcements.Reinforce(OrdosMain, InitialOrdosReinforcements[1], InitialOrdosPaths[1]), Reinforcements.Reinforce(OrdosMain, InitialOrdosReinforcements[2], InitialOrdosPaths[2]))
 	IdlingUnits[OrdosSmall] = Reinforcements.Reinforce(OrdosSmall, InitialOrdosReinforcements[1], InitialOrdosPaths[3])
 
+
```
newline

pchote commented on mods/d2k/maps/harkonnen-06b/harkonnen06b.lua:
```diff
@@ -118,6 +125,52 @@ SendStarportReinforcements = function()
+		 end
+	end)
+	end
+
```
newline

pchote commented on mods/d2k/maps/harkonnen-06b/harkonnen06b.lua:
```diff
@@ -118,6 +125,52 @@ SendStarportReinforcements = function()
+	unit.Stop()
+	unit.AttackMove(intruder.Location,1)
+	Trigger.AfterDelay(100, function() FollowIntruder(unit, intruder) end)
+
```
newline

pchote commented on mods/d2k/maps/harkonnen-07/harkonnen07-AI.lua:
```diff
@@ -14,29 +35,54 @@ AttackGroupSize =
+	hard = { DateTime.Seconds(2), DateTime.Seconds(4) }
 }
 
+
```
newline

pchote commented on mods/d2k/maps/harkonnen-07/harkonnen07.lua:
```diff
@@ -147,11 +147,34 @@ SendEnemyReinforcements = function(player, delay, path, unitTypes, customConditi
+	if Utils.RandomInteger(1, 100) < 30 then
+		Trigger.AfterDelay(1000, function() AirstrikeLogic(airstrikeProvider)end)
+	else
+		Media.Debug("Airstrike vs building")
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-07/harkonnen07.lua:
```diff
@@ -163,12 +186,39 @@ SendAirStrike = function()
+	if airstrikeProvider.IsDead or DateTime.GameTime <= AirStrikeTimer then return end
+	local bestValue = {}
+	local bestIndex = 1
+	Media.Debug("possible target count: "..tostring(#possibleTargets))
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-07/harkonnen07.lua:
```diff
@@ -163,12 +186,39 @@ SendAirStrike = function()
+			bestIndex = i
+		end
+	end
+		Media.Debug("selected target is "..tostring(possibleTargets[bestIndex]).."value: "..bestValue[bestIndex])
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-07/harkonnen07.lua:
```diff
@@ -163,12 +186,39 @@ SendAirStrike = function()
+		AirStrikeTimer =  DateTime.GameTime + AirStrikeChargeTime
 end
 
+
```
newlines

pchote commented on mods/d2k/maps/harkonnen-07/harkonnen07.lua:
```diff
@@ -212,14 +262,37 @@ Tick = function()
 		end
 	end
+	if IdlingUnits[AtreidesMain] == nil then return end
+	UserInterface.SetMissionText("MainBase:      "..tostring(#IdlingUnits[AtreidesMain]).."      Cash:"..tostring(AtreidesMain.Cash+AtreidesMain.Resources).."\n".."Smallbase:      "..tostring(#IdlingUnits[AtreidesSmall]).."      Cash:"..tostring(AtreidesSmall.Cash+AtreidesSmall.Resources))
```
String needs to be extracted to fluent.

pchote commented on mods/d2k/maps/harkonnen-07/harkonnen07.lua:
```diff
@@ -212,14 +262,37 @@ Tick = function()
+end
+
+EmergencyBehaviour = function(player, target)
+	Media.Debug("airstrike close support")
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-07/harkonnen07.lua:
```diff
@@ -212,14 +262,37 @@ Tick = function()
+				and a.HasProperty("Attack")
+	end)
+	if targets[1] == nil  then return end
+	Media.Debug("sending"..tostring(#targets))
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-08/harkonnen08-AI.lua:
```diff
@@ -44,26 +112,101 @@ ActivateAI = function()
+	}
+	Trigger.OnBuildingPlaced(Ordos, function(p, building)
+		table.insert(OrdosBase, building)
+		Media.Debug("found building : "..tostring(building))
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-08/harkonnen08-AI.lua:
```diff
@@ -44,26 +112,101 @@ ActivateAI = function()
+		Media.Debug("found building : "..tostring(building))
+		DefendAndRepairBase(Ordos, {building},0.5, AttackGroupSize[Difficulty] )
+		if productionTypesOrdos[building.Type] == nil then return end
+		Media.Debug("Restart production: "..tostring(building.Type))
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-08/harkonnen08.lua:
```diff
@@ -133,10 +158,34 @@ SendAirStrike = function()
+	if airstrikeProvider.IsDead or DateTime.GameTime <= AirStrikeTimer then return end
+	local bestValue = {}
+	local bestIndex = 1
+	Media.Debug("possible target count: "..tostring(#possibleTargets))
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-08/harkonnen08.lua:
```diff
@@ -133,10 +158,34 @@ SendAirStrike = function()
+			bestIndex = i
+		end
+	end
+		Media.Debug("selected target is "..tostring(possibleTargets[bestIndex]).."value: "..bestValue[bestIndex])
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-08/harkonnen08.lua:
```diff
@@ -133,10 +158,34 @@ SendAirStrike = function()
+	end
+		Media.Debug("selected target is "..tostring(possibleTargets[bestIndex]).."value: "..bestValue[bestIndex])
+		airstrikeProvider.TargetAirstrike(possibleTargets[bestIndex].CenterPosition)
+		AirStrikeTimer =  DateTime.GameTime + AirStrikeChargeTime
```
Remove double space

pchote commented on mods/d2k/maps/harkonnen-08/harkonnen08.lua:
```diff
@@ -162,16 +211,58 @@ BuildSaboteur = function()
+		end)
+		return
+	end
+	-- filter out targets where infatry is nearby
```
typo "infatry"

pchote commented on mods/d2k/maps/harkonnen-08/harkonnen08.lua:
```diff
@@ -162,16 +211,58 @@ BuildSaboteur = function()
+		local dfd = Utils.Random(possibleTargets)
+		Media.Debug("demolising "..tostring(dfd))
+		saboteur.Demolish(dfd)
+		--unsucceffull
```
typo

pchote commented on mods/d2k/maps/harkonnen-08/harkonnen08.lua:
```diff
@@ -216,6 +307,23 @@ ChangeOwner = function(old_owner, new_owner)
+EmergencyBehaviour = function(player,target)
+	HoldProduction[player] = false
+	if player == Ordos and not AtreidesEnemy.HasNoRequiredUnits() then
+		Media.Debug("airstrike close support")
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-08/harkonnen08.lua:
```diff
@@ -216,6 +307,23 @@ ChangeOwner = function(old_owner, new_owner)
+					and a.HasProperty("Attack")
+		end)
+		if enemyunits[1] == nil  then return end
+		Media.Debug("possieble targets for airtrike: "..tostring(#enemyunits))
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-08/harkonnen08.lua:
```diff
@@ -254,8 +362,12 @@ Tick = function()
 		end
 	end
+	if DateTime.GameTime % DateTime.Seconds(1) then
+		DebugInfo()
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-09a/harkonnen09a-AI.lua:
```diff
@@ -7,33 +7,115 @@
-AttackDelays =
+EarlyProductionDelays =
+{
+
```
stray newline

pchote commented on mods/d2k/maps/harkonnen-09a/harkonnen09a-AI.lua:
```diff
@@ -7,33 +7,115 @@
+	DefencePerimeter[AtreidesSmall2] = GetCellsInRectangle(CPos.New(18,32), CPos.New(30,37))
+	DefencePerimeter[CorrinoMain] = GetCellsInRectangle(CPos.New(3,13), CPos.New(24,26))
+	DefencePerimeter[CorrinoSmall] = GetCellsInRectangle(CPos.New(74,27), CPos.New(96,48))
+
```
newline

pchote commented on mods/d2k/maps/harkonnen-09a/harkonnen09a-AI.lua:
```diff
@@ -46,35 +128,108 @@ ActivateAI = function()
+	end
 	local infantryToBuild = function() return { Utils.Random(EnemyInfantryTypes) } end
 	local infantryToBuildCorrinoMain = function() return { Utils.Random(CorrinoMainInfantryTypes) } end
 	local vehilcesToBuild = function() return { Utils.Random(EnemyVehicleTypes) } end
```
typo

pchote commented on mods/d2k/maps/harkonnen-07/harkonnen07-AI.lua:
```diff
@@ -45,23 +91,70 @@ ActivateAI = function()
-		ProduceUnits(Corrino, CBarracks, delay, infantryToBuildCorrino, AttackGroupSize[Difficulty], attackThresholdSize)
+	Trigger.OnBuildingPlaced(AtreidesMain, function(p, building)
+		table.insert(AtreidesMainBase, building)
+		DefendAndRepairBase(AtreidesMain, {building},0.5, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-08/harkonnen08-AI.lua:
```diff
@@ -44,26 +112,101 @@ ActivateAI = function()
+	Trigger.OnBuildingPlaced(Ordos, function(p, building)
+		table.insert(OrdosBase, building)
+		Media.Debug("found building : "..tostring(building))
+		DefendAndRepairBase(Ordos, {building},0.5, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-09a/harkonnen09a-AI.lua:
```diff
@@ -46,35 +128,108 @@ ActivateAI = function()
-	ProduceUnits(CorrinoSmall, CHeavyFactory2, delay, tanksToBuildCorrinoSmall, AttackGroupSize[Difficulty], attackThresholdSize)
+	Trigger.OnBuildingPlaced(AtreidesMain, function(p, building)
+		table.insert(AtreidesMainBase, building)
+		DefendAndRepairBase(AtreidesMain, {building},0.5, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-09a/harkonnen09a-AI.lua:
```diff
@@ -46,35 +128,108 @@ ActivateAI = function()
+	end)
+	Trigger.OnBuildingPlaced(AtreidesSmall1, function(p, building)
+		table.insert(AtreidesSmallBase, building)
+		DefendAndRepairBase(AtreidesSmall1, {building},0.5, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-09a/harkonnen09a-AI.lua:
```diff
@@ -46,35 +128,108 @@ ActivateAI = function()
+	end)
+	Trigger.OnBuildingPlaced(CorrinoMain, function(p, building)
+		table.insert(CorrinoMainBase, building)
+		DefendAndRepairBase(CorrinoMain, {building},0.5, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-09a/harkonnen09a.lua:
```diff
@@ -131,9 +131,14 @@ InitialCorrinoPaths =
-HarkonnenReinforcements = { "combat_tank_h", "combat_tank_h", "siege_tank", "siege_tank", "missile_tank" }
+HarkonnenReinforcements = {
+	{ "combat_tank_h", "combat_tank_h", "siege_tank", "siege_tank", "missile_tank" },
+	{"trooper", "light_inf", "light_inf", "light_inf", "trooper", "light_inf", "trooper", "light_inf", "trooper", "light_inf", "trooper" },
```
missing spaces at start of lists.

pchote commented on mods/d2k/maps/harkonnen-09a/harkonnen09a.lua:
```diff
@@ -178,12 +216,88 @@ SendAirStrike = function()
+	if airstrikeProvider.IsDead or DateTime.GameTime <= AirStrikeTimer then return end
+	local bestValue = {}
+	local bestIndex = 1
+	Media.Debug("possible target count: "..tostring(#possibleTargets))
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-09a/harkonnen09a.lua:
```diff
@@ -178,12 +216,88 @@ SendAirStrike = function()
+			bestIndex = i
+		end
+	end
+		Media.Debug("selected target is "..tostring(possibleTargets[bestIndex]).."value: "..bestValue[bestIndex])
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-09a/harkonnen09a.lua:
```diff
@@ -178,12 +216,88 @@ SendAirStrike = function()
+
+EmergencyBehaviour = function(player,target)
+	if player == AtreidesMain or player == AtreidesSmall1 then
+		Media.Debug("airstrike close support")
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-09a/harkonnen09a.lua:
```diff
@@ -178,12 +216,88 @@ SendAirStrike = function()
+					and a.HasProperty("Attack")
+		end)
+		if enemyunits[1] == nil  then return end
+		Media.Debug("possieble targets for airtrike: "..tostring(#enemyunits))
```
remove debug?

pchote commented on mods/d2k/maps/harkonnen-09b/harkonnen09b-AI.lua:
```diff
@@ -7,32 +7,113 @@
+	EarlyGame = { "combat_tank_h" },
+	LateGame = { "combat_tank_h", "combat_tank_h", "siege_tank", "missile_tank" }
+}
+
```
newline

pchote commented on mods/d2k/maps/harkonnen-09b/harkonnen09b-AI.lua:
```diff
@@ -7,32 +7,113 @@
+	DefencePerimeter[CorrinoMain] = GetCellsInRectangle(CPos.New(65,38), CPos.New(91,55))
+	DefencePerimeter[CorrinoSmall] = GetCellsInRectangle(CPos.New(33,35), CPos.New(49,52))
+	Media.Debug("First: "..tostring(DefencePerimeter[CorrinoSmall][1]).."last: "..tostring(DefencePerimeter[CorrinoSmall][#DefencePerimeter[CorrinoSmall]]))
+
```
newline

pchote commented on mods/d2k/maps/harkonnen-09b/harkonnen09b-AI.lua:
```diff
@@ -42,32 +123,101 @@ ActivateAI = function()
+			return { Utils.Random(CorrinoTankTypes["LateGame"]) }
+		end
+	end
+
```
newline

pchote commented on mods/d2k/maps/harkonnen-09b/harkonnen09b-AI.lua:
```diff
@@ -42,32 +123,101 @@ ActivateAI = function()
+	local attackThresholdSize = AttackGroupSize[Difficulty] * 3
+	local attackThresholdSizeSmall = AttackGroupSize[Difficulty] * 2
+	local attackThresholdSizeEmperor = AttackGroupSize[Difficulty] * 5
+
```
newline

pchote commented on mods/d2k/maps/harkonnen-09b/harkonnen09b-AI.lua:
```diff
@@ -42,32 +123,101 @@ ActivateAI = function()
+	end
+	Trigger.OnBuildingPlaced(AtreidesMain, function(p, building)
+		table.insert(AtreidesMainBase, building)
+		DefendAndRepairBase(AtreidesMain, {building},0.5, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-09b/harkonnen09b-AI.lua:
```diff
@@ -42,32 +123,101 @@ ActivateAI = function()
+	end)
+	Trigger.OnBuildingPlaced(AtreidesSmall, function(p, building)
+		table.insert(AtreidesSmallBase, building)
+		DefendAndRepairBase(AtreidesSmall, {building},0.5, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-09b/harkonnen09b-AI.lua:
```diff
@@ -42,32 +123,101 @@ ActivateAI = function()
+	end)
+	Trigger.OnBuildingPlaced(CorrinoMain, function(p, building)
+		table.insert(CorrinoMainBase, building)
+		DefendAndRepairBase(CorrinoMain, {building},0.5, AttackGroupSize[Difficulty] )
```
missing space

pchote commented on mods/d2k/maps/harkonnen-09b/harkonnen09b.lua:
```diff
@@ -167,12 +167,34 @@ SendHarkonnenReinforcements = function(delay)
+	if Utils.RandomInteger(1, 100) < 30 then
+		Trigger.AfterDelay(1501, function() AirstrikeLogic(airstrikeProvider)end)
+	else
+		Media.Debug("Airstrike vs building")
```
debug

pchote commented on mods/d2k/maps/harkonnen-09b/harkonnen09b.lua:
```diff
@@ -184,17 +206,76 @@ SendAirStrike = function()
+	if airstrikeProvider.IsDead then return end
+	local bestValue = {}
+	local bestIndex = 1
+	Media.Debug("possible target count: "..tostring(#possibleTargets))
```
debug

pchote commented on mods/d2k/maps/harkonnen-09b/harkonnen09b.lua:
```diff
@@ -184,17 +206,76 @@ SendAirStrike = function()
+			bestIndex = i
+		end
+	end
+		Media.Debug("selected target is "..tostring(possibleTargets[bestIndex]).."value: "..bestValue[bestIndex])
```
debug

pchote commented on mods/d2k/maps/harkonnen-09b/harkonnen09b.lua:
```diff
@@ -184,17 +206,76 @@ SendAirStrike = function()
+
+EmergencyBehaviour = function(player,target)
+	if player == AtreidesMain or player == AtreidesSmall then
+		Media.Debug("airstrike close support")
```
debug

pchote commented on mods/d2k/maps/harkonnen-09b/harkonnen09b.lua:
```diff
@@ -184,17 +206,76 @@ SendAirStrike = function()
+					and a.HasProperty("Attack")
+		end)
+		if enemyunits[1] == nil  then return end
+		Media.Debug("possieble targets for airtrike: "..tostring(#enemyunits))
```
debug

pchote commented on mods/d2k/maps/ordos-03a/ordos03a-AI.lua:
```diff
@@ -24,7 +24,15 @@ AttackDelays =
 HarkonnenInfantryTypes = { "light_inf", "light_inf", "light_inf", "trooper", "trooper" }
 HarkonnenVehicleTypes = { "trike", "trike", "quad" }
 
+
```
newlines

pchote commented on mods/d2k/maps/ordos-03b/ordos03b-AI.lua:
```diff
@@ -24,20 +24,33 @@ AttackDelays =
 HarkonnenInfantryTypes = { "light_inf", "light_inf", "light_inf", "trooper", "trooper" }
 HarkonnenVehicleTypes = { "trike", "trike", "quad" }
 
+
```
newline

pchote commented on mods/d2k/maps/ordos-05/ordos05-AI.lua:
```diff
@@ -25,6 +51,7 @@ EnemyInfantryTypes = { "light_inf", "light_inf", "light_inf", "trooper", "troope
 EnemyVehicleTypes = { "trike", "trike", "quad" }
 EnemyTankType = { "combat_tank_a" }
 
+
```
newline

pchote commented on mods/d2k/maps/ordos-06a/ordos06a-AI.lua:
```diff
@@ -27,6 +47,7 @@ EnemyVehicleTypes = { "trike", "trike", "quad" }
 AtreidesTankTypes = { "combat_tank_a", "siege_tank", "missile_tank" }
 HarkonnenTankTypes = { "combat_tank_h", "siege_tank" }
 
+
```
newline

pchote commented on mods/d2k/rules/campaign-rules.yaml:
```diff
@@ -19,9 +19,16 @@ Player:
 	ModularBot@CampaignAI:
 		Name: bot-campaign-ai.name
 		Type: campaign
+	GrantConditionOnBotOwner@campain:
```
typo

pchote commented on mods/d2k/rules/campaign-rules.yaml:
```diff
@@ -65,5 +72,443 @@ sandworm:
+	Targetable:
+		TargetTypes: Ground, Vehicle, Husk
+		RequiresForceFire: true
+
```
newline

pchote commented on mods/d2k/rules/campaign-rules.yaml:
```diff
@@ -65,5 +72,443 @@ sandworm:
+	SpawnActorOnDeath@debris1:
+		Actor: debris.large
+		OwnerType: InternalName
+
```
newline

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -108,14 +134,62 @@ end
+---@type table<player, boolean>
+EmergencyBuildRate = {}
+
+
```
newlines

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -108,14 +134,62 @@ end
+--- List of all unit combat types.
+AllAttackUnitTypes = { "light_inf", "trooper", "trike", "quad", "siege_tank", "missile_tank", "sonic_tank", "devastator", "raider", "stealth_raider", "deviator", "combat_tank_a", "combat_tank_h", "combat_tank_o" }
+
+
```
newline

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -108,14 +134,62 @@ end
+
+
+--- Activates when EmergencyBuildRate is True.
+--- Overide this functions for mission specific behaviour
```
typos `Overide`, `functions`

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -126,15 +200,16 @@ LastHarvesterEaten = { }
 SetupAttackGroup = function(owner, size)
 	local units = { }
-
+	-- IdlingUnits can contain already dead actors or eated actors by worm
```
can -> may.

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -150,53 +225,98 @@ SendAttack = function(owner, size)
 	if Attacking[owner] then
 		return
 	end
+	Media.Debug("attack started")
```
remove debug

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -150,53 +225,98 @@ SendAttack = function(owner, size)
+		end)
+	end)
 	Trigger.OnAllRemovedFromWorld(units, function()
+		Media.Debug("attack stoped")
```
remove debug

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -150,53 +225,98 @@ SendAttack = function(owner, size)
+	if PatrolPoints[owner] == nil then return end
+	local units = SetupAttackGroup(owner, size)
+	local cells = Utils.Shuffle(PatrolPoints[owner])
+	Media.Debug("sending patrol")
```
remove debug

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -150,53 +225,98 @@ SendAttack = function(owner, size)
+			unit.Patrol(cells, false, 1000)
+			unit.CallFunc(function()
+					table.insert(IdlingUnits[owner], unit)
+					Media.Debug("Patrol ends "..tostring(unit))
```
emove debug

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +368,253 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+				if EmergencyBuildRate[owner] then return end
+				EmergencyBuildRate[owner] = true
+				EmergencyBehaviour(owner, killer.Location)
+				Media.Debug("Emergency - building destroyed")
```
remove debug

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +368,253 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
-
+	--- Check harvester count
+	if factory.Type == "heavy_factory" and #player.GetActorsByType("harvester") < HarvesterCount[player] then
+		Media.Debug("Production overide: harvester for"..tostring(player.InternalName))
```
remove debug

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +368,253 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
 		if #IdlingUnits[player] >= attackThresholdSize then
-			SendAttack(player, attackSize)
+			if DateTime.GameTime < AttackDelay[player] then
+				Media.Debug(tostring(player.InternalName).." too early  sending patrol - holding production for"..tostring(AttackDelay[player] - DateTime.GameTime ))
```
remove debug

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +368,253 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+	end
+end
+
+
```
newlines

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +368,253 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+		end
+	end)
+end
+
```
newline

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +368,253 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+--- @param size? integer (optional). When nil random integer is selected
+CheckArea = function(owner, targetArea, size)
+	if #IdlingUnits[owner] == 0 then
+		--Media.Debug("no available units")
```
remove commented debug

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +368,253 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+		--Media.Debug("no available units")
+		return
+	end
+	Media.Debug("checking area "..tostring(targetArea))
```
debug

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +368,253 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+				and a.Owner.InternalName ~= "Neutral"
+				and a.Owner.InternalName ~= "Creeps"
+		end)
+		--Media.Debug("actor: "..tostring(unit).." Find targets "..tostring(#enemies))
```
commented debug

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +368,253 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+			unit.CallFunc(function()
+				if not unit.IsDead then
+					table.insert(IdlingUnits[owner], unit)
+					--Media.Debug("Return for routine "..tostring(unit))
```
commented debug

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +368,253 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+			local cell = CPos.New(x, y)
+			if Map.TerrainType(cell) ~= "Cliff" and Map.TerrainType(cell) ~= "Rough" then
+				cells[index] = cell
+				--Media.Debug("cell is"..tostring(cells[index]))
```
commented debug

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +368,253 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+RemoveDeadActors = function (actors)
+	for i = #actors, 1, -1 do
+		if actors[i].IsDead or not actors[i].IsInWorld then
+			Media.Debug("removing dead actor")
```
debug

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +368,253 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+RemoveActor = function (actors, actor)
+	for i = #actors, 1, -1 do
+		if actors[i] == actor then
+			--Media.Debug(" actor removed "..tostring(actor))
```
commented debug

pchote commented on mods/d2k/scripts/campaign.lua:
```diff
@@ -248,32 +368,253 @@ DefendAndRepairBase = function(owner, baseBuildings, modifier, defenderCount)
+	end
+end
+
+DebugInfo = function()
```
remove debug / many typos.

pchote reviewed (commented):
I don't see anything particularly concerning in the updated scripts, and some of the new script features look great.

Commented many many typo/formatting nits, plus a couple of hardcoded strings that need to move to Fluent.

I volunteer to fix all of these over the next few days.

pchote commented:
I believe i've resolved all the style/typo issues.

JovialFeline commented on mods/d2k/rules/campaign-rules.yaml:
```diff
@@ -19,9 +19,16 @@ Player:
+	GrantConditionOnBotOwner@campaign:
+		Condition: campaign-ai
+		Bots: campaign
+	ProvidesPrerequisite@autoConreteForBots:
```
```suggestion
	ProvidesPrerequisite@autoConcreteForBots:
```

Porenutak committed:
Rework D2k campaign scripts:

- Add InitialProductionDelay
- Different attacksDelays for early game and late game
- Adjust T3 unit production for selected missions
- Implement InitialAttackDelay, TimeBetweenAttacks
- Implement routines for idle units
- Emergency mode
- Airstrike rework

pchote commented:
Fixed.

PunkPun reviewed (approved):
alright, let's itterate on the playtest

PunkPun commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/8842d9c53224a7bd8ac031fdf2a19b293fac7056...c4462d92d3f8c80be7ee14e1b99e8ec00e7929a8
